astra/src/main/java/com/slack/astra/clusterManager/CacheNodeSearchabilityService.java

```diff
@@ -0,0 +1,129 @@
+  private final SnapshotMetadataStore snapshotMetadataStore;
+  private final AstraConfigs.ManagerConfig managerConfig;
+  private final Counter numberOfUnsearchableNodes;
+
+  public CacheNodeSearchabilityService(
+      MeterRegistry meterRegistry,
+      CacheNodeMetadataStore cacheNodeMetadataStore,
+      AstraConfigs.ManagerConfig managerConfig,
+      CacheNodeAssignmentStore cacheNodeAssignmentStore,
+      SearchMetadataStore searchMetadataStore,
+      SnapshotMetadataStore snapshotMetadataStore) {
+    this.cacheNodeMetadataStore = cacheNodeMetadataStore;
+    this.managerConfig = managerConfig;
+    this.cacheNodeAssignmentStore = cacheNodeAssignmentStore;
+    this.searchMetadataStore = searchMetadataStore;
+    this.snapshotMetadataStore = snapshotMetadataStore;
+    this.numberOfUnsearchableNodes = meterRegistry.counter(UNSEARCHABLE_NODES);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closing cache node searchability service");
+    if (cacheNodeMetadataStore != null) {
+      cacheNodeMetadataStore.close();
+    }
+    if (cacheNodeAssignmentStore != null) {
+      cacheNodeAssignmentStore.close();
+    }
+    if (searchMetadataStore != null) {
+      searchMetadataStore.close();
+    }
+    if (snapshotMetadataStore != null) {
+      snapshotMetadataStore.close();
+    }
+    LOG.info("Closed cache node assignment service");
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    List<CacheNodeMetadata> unsearchableCacheNodes =
+        cacheNodeMetadataStore.listSync().stream()
+            .filter(cacheNodeMetadata -> !cacheNodeMetadata.searchable)
+            .toList();
+
+    for (CacheNodeMetadata cacheNodeMetadata : unsearchableCacheNodes) {
+      List<CacheNodeAssignment> loadingCacheAssignments = new ArrayList<>();
+      List<CacheNodeAssignment> liveCacheAssignments = new ArrayList<>();
+      cacheNodeAssignmentStore
+          .listSync()
+          .forEach(
+              cacheNodeAssignment -> {
+                if (cacheNodeMetadata.id.equals(cacheNodeAssignment.cacheNodeId)) {
+                  if (cacheNodeAssignment.state
+                      == Metadata.CacheNodeAssignment.CacheNodeAssignmentState.LOADING) {
+                    loadingCacheAssignments.add(cacheNodeAssignment);
+                  } else if (cacheNodeAssignment.state
+                      == Metadata.CacheNodeAssignment.CacheNodeAssignmentState.LIVE) {
+                    liveCacheAssignments.add(cacheNodeAssignment);
+                  }
+                }
+              });
+
+      // This node is only searchable if it doesn't have more than 1 assignment that is
+      // loading AND it has at least one assignment that is live
+      boolean searchable = loadingCacheAssignments.size() <= 1 && !liveCacheAssignments.isEmpty();
+
+      if (searchable) {
+        LOG.info("Marking cache node {} as searchable", cacheNodeMetadata.id);
```

Comment:
Is the `id` useful in a log message here? Can we identify it without looking at the code?